internal/ethapi/api.go

```diff
@@ -583,10 +583,15 @@ type CallArgs struct {
 func (s *PublicBlockChainAPI) doCall(ctx context.Context, args CallArgs, blockNr rpc.BlockNumber, vmCfg vm.Config) ([]byte, uint64, bool, error) {
 	defer func(start time.Time) { log.Debug("Executing EVM call finished", "runtime", time.Since(start)) }(time.Now())
 
-	state, header, err := s.b.StateAndHeaderByNumber(ctx, blockNr)
+	state, err := s.b.StateByNumber(ctx, blockNr)
 	if state == nil || err != nil {
 		return nil, 0, false, err
 	}
+	header, err := s.b.HeaderByNumber(ctx, blockNr)
+	if header == nil || err != nil {
+		return nil, 0, false, err
+	}
```

Comment:
This is racey. The whole reason behind retrieving the header and the state at the same time is because the pending state can and will change very frequently, often between these two calls (retrieving the state and the header separately). This was our original solution too btw that we needed to fix, took quite a lot to detect and figure out the issue.